docs/src/main/sphinx/develop/connectors.rst

```diff
@@ -79,33 +79,102 @@ features, like:
 
+When a ``SELECT`` query is being executed, and it contains specific clauses,
+the engine can try pushing down some of them into the connector, by calling
+one of the following methods of the ``ConnectorMetadata`` service:
+
+* ``applyLimit`` for the ``LIMIT`` clause, or ``applyTopN`` when ``ORDER BY``
+  is also present
+* ``applyFilter`` for the ``WHERE`` clause
+* ``applyProjection`` for expressions in the ``SELECT`` clause
+* ``applySample`` for the ``TABLESAMPLE`` clause
+* ``applyAggregation`` for the ``GROUP BY`` clause
+* ``applyJoin`` for the different types of the ``JOIN`` clause
+* ``applyTableFunction`` for the ``TABLE ()`` clause
+* ``applyTableScanRedirect``
+
+
+Connectors can indicate whether they don't support a particular pushdown
+or that the action had no effect by returning ``Optional.empty()``. Connectors
+should expect these methods to be called multiple times during the optimization
+of a given query.
+
+.. warning::
+
+  It's critical for connectors to return ``Optional.empty()`` if calling
+  this method has no effect for that invocation, even if the connector generally
+  supports a particular pushdown. Doing otherwise can cause the optimizer
+  to loop indefinitely.
+
+Otherwise, these methods return a result object containing the connectors
+table handle. The connector can store any kind of information in the table
+handle, as it will be passed later to other services it implements, like
+the record set or page source provider.
+
+Limit and TopN pushdown
+^^^^^^^^^^^^^^^^^^^^^^^
+
+When a ``SELECT`` query contains a :ref:`limit-clause`, the engine will try to
+push down the limit into the connector. If there's no ``ORDER BY`` clause,
+the ``applyLimit`` method of the connector metadata service will be called,
+and the connector can return results in an arbitrary order.
+Otherwise, the engine will call ``applyTopN``.
+
+If the connector could benefit from the information but can't guarantee that it
+will be able to produce fewer rows than the provided limit, it should return
+a non-empty result containing a new handle for the derived table and
+the "limit guaranteed" or "topN guaranteed" flag set to false.
+
+If the connector can guarantee it will produce fewer rows than the provided
+limit, it should return a non-empty result with the "limit guaranteed" or
+"topN guaranteed" flag set to true.
+
+.. note::
+
+  The ``applyTopN`` is the only method that receives sort items from the
+  ``ORDER BY`` clause. It will be called even if there's no ``LIMIT`` clause.
+
+In an SQL query, the ``ORDER BY`` section can include any column with any order.
+But the data source for the connector might only support limited combinations.
+Plugin authors have to decide if the connector should ignore the pushdown,
+return all the data and let the engine sort it, or throw an exception
+to inform the user that particular order is not supported, if fetching all
+the data would be too expensive or time consuming. When throwing
```

Comment:
Do we have any connectors which already do this today?

Reply:
I do, https://github.com/nineinchnick/trino-rest

Reply:
Thanks. However all error codes in StandardErrorCodes are supposed to be thrown by engine. Connectors should define their own. At-least that's my understanding from some previous PRs.

cc: @martint would know better.

Reply:
Sure, I can describe how to create custom exceptions. I was planning on adding a whole paragraph/section about error handling but in another PR.

Reply:
@martint WDYT?

Reply:
> However all error codes in StandardErrorCodes are supposed to be thrown by engine

If an error is covered by one of the entries in StandardErrorCodes, connectors should use it. They should only define custom codes for non-standard (i.e., connector-specific) errors.

Reply:
So I'll leave it as is. We'll have more opportunities to mention connector-specific errors.